vllm/model_executor/layers/fused_moe/layer.py

```diff
@@ -859,23 +884,46 @@ def forward_impl(self, hidden_states: torch.Tensor,
                                                  cu_tokens_across_dp_cpu)
 
         # Matrix multiply.
-        final_hidden_states = self.quant_method.apply(
-            layer=self,
-            x=hidden_states,
-            router_logits=router_logits,
-            top_k=self.top_k,
-            renormalize=self.renormalize,
-            use_grouped_topk=self.use_grouped_topk,
-            global_num_experts=self.global_num_experts,
-            expert_map=self.expert_map,
-            topk_group=self.topk_group,
-            num_expert_group=self.num_expert_group,
-            custom_routing_function=self.custom_routing_function,
-            scoring_func=self.scoring_func,
-            e_score_correction_bias=self.e_score_correction_bias,
-            activation=self.activation,
-            apply_router_weight_on_input=self.apply_router_weight_on_input,
-        )
+        from vllm.model_executor.layers.quantization.fp8 import Fp8MoEMethod
+        if self.num_share_fusion_replicas > 0 and type(self.quant_method) in [
+                Fp8MoEMethod, UnquantizedFusedMoEMethod
+        ]:
```

Comment:
This is making me a little nervous. Do we only support shared expert fusion for these two quant methods? What's left to do to support this generally? If it's easy (i.e. just plumbing) I think we should take care of it before landing this PR.

If we don't support this generally, it adds complexity to the support matrix and makes things harder to maintain/develop.